[Wio 3G] Adding platform

ytsuboi opened the pull request:
### Description

Adding new platform [Wio 3G](https://os.mbed.com/platforms/Seeed-Wio-3g/), STM32F439VI with Quectel UG96 on-board 3G modem.

I don't want to use preprocessor macro to switch AT command. But since member function of cellular feature such as set_context_to_be_activated() are private, I decided to use macro to reduce shared code changes ( especially changing code structure ).

Test results:  
[GCC](https://github.com/ARMmbed/mbed-os/files/2070858/Mbed-Wio_3G-testresult.txt) 
[armcc](https://github.com/ARMmbed/mbed-os/files/2070860/Mbed-Wio_3G-armcc-testresult.txt) 
[EWARM](https://github.com/ARMmbed/mbed-os/files/2070862/Mbed-Wio_3G-IAR-testresult.txt)




### Pull request type

    [ ] Fix
    [ ] Refactor
    [x] New target
    [ ] Feature
    [ ] Breaking change

ytsuboi committed:
[Wio 3G] Adding platform

0xc0170 commented:
@ytsuboi only GCC ARM test results provided?

ytsuboi commented:
@0xc0170 Added on description.
@toyowata Thanks for your help!

ashok-rao commented on targets/targets.json:
```diff
@@ -1724,6 +1724,40 @@
+        "config": {
+            "clock_source": {
+                "help": "Mask value : USE_PLL_HSE_EXTC | USE_PLL_HSE_XTAL (need HW patch) | USE_PLL_HSI",
+                "value": "USE_PLL_HSE_XTAL|USE_PLL_HSI",
```
Question: What is the HW patch? Is this explained somewhere in the docs?

ytsuboi replied:
Oops, no need hardware patch. XTAL is mounted on the board. Will remove "need HW patch" on next commit.

ytsuboi replied:
Done

ashok-rao commented on targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/TARGET_WIO_3G/PinNames.h:
```diff
@@ -0,0 +1,268 @@
+    USB_OTG_HS_ULPI_DIR = PC_2,
+    USB_OTG_HS_ULPI_NXT = PC_3,
+*/
+
```
I think all of these un-used USB pins can be removed..

ytsuboi replied:
Done

ashok-rao commented on targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/TARGET_WIO_3G/PinNames.h:
```diff
@@ -0,0 +1,268 @@
+    LED_RED     = LED2,
+    USER_BUTTON = PC_13,
+    // Standardized button names
+//    BUTTON1 = USER_BUTTON,
```
This one is required for consistency.

ytsuboi replied:
Done

ashok-rao reviewed (dismissed):
LGTM. Thanks, nice work!

0xc0170 commented:
@ARMmbed/mbed-os-wan Can you review cellular additions here?

AnttiKauppila commented on features/cellular/framework/AT/AT_CellularNetwork.cpp:
```diff
@@ -429,11 +429,21 @@ nsapi_error_t AT_CellularNetwork::set_context_to_be_activated()
 
     // if user has defined user name and password we need to call CGAUTH before activating or modifying context
     if (_pwd && _uname) {
+#if defined(TARGET_WIO_3G)
```
This is not a correct approach. Now I understand the context of your earlier question about  set_context_to_be_activated accessibility. I agree it does not make sense to copy paste activate_context method to your own class.
Please propose a solution where set_context_to_be_activated is changed to protected and virtual method, so you can override that in your class.

mirelachirica replied:
What about introducing new protected virtual method for "user authentication" only and override that one?

jarvte replied:
I would like @mirelachirica approach, clear method for authentication. Does not need to be in API level.

ytsuboi replied:
I can solve this issue with @AnttiKauppila approach soon, but @mirelachirica approach is too heavy for me. ( I don't want to change structures of shared/common code, since it could affect others platform. ) In addition, I don't want to support separating a function from AT command to response.

So, can I change set_context_to_be_activated() to protected and virtual method? @mirelachirica @jarvte

mirelachirica replied:
I don't see any risk in changing the shared code something like this:

```
nsapi_error_t AT_CellularNetwork::set_context_to_be_activated()
{
    // try to find or create context with suitable stack
    if (!get_context()) {
        return NSAPI_ERROR_NO_CONNECTION;
    }

    // if user has defined user name and password we need to call CGAUTH before activating or modifying context
    nsapi_error_t ret_val = do_user_authentication();
    
    return ret_val;
}

nsapi_error_t AT_CellularNetwork::do_user_authentication()
{
    if (_pwd && _uname) {
        _at.cmd_start("AT+CGAUTH=");
        _at.write_int(_cid);
        _at.write_int(_authentication_type);
        _at.write_string(_uname);
        _at.write_string(_pwd);
        _at.cmd_stop();
        _at.resp_start();
        _at.resp_stop();
        if (_at.get_last_error() != NSAPI_ERROR_OK) {
            return NSAPI_ERROR_AUTH_FAILURE;
        }
    }
    return _at.get_last_error();
}
```
```

AnttiKauppila replied:
But with this approach, what is the need of set_context_to_be_activated function?
Also calling inheritable function from private method is not a good design.
I think the best would be to rename set_context_to_be_activated() to do_user_authentication()
We can do this without breaking anything as it was a private method.
Also is get_context() check really needed in this function or should it be in a place where this function is called (activate_context)? Usually private functions are only called when the internal status is ok. And checking what activate_context does it makes perfect sense to move this check there.

mirelachirica replied:
I agree with @AnttiKauppila proposals. One note still: not only rename set_context_to_be_activated() to do_user_authentication() but make it also protected virtual.

And the usage like:

```
nsapi_error_t AT_CellularNetwork::do_user_authentication()
{
    if (_pwd && _uname) {
        _at.cmd_start("AT+CGAUTH=");
        _at.write_int(_cid);
        _at.write_int(_authentication_type);
        _at.write_string(_uname);
        _at.write_string(_pwd);
        _at.cmd_stop();
        _at.resp_start();
        _at.resp_stop();
        if (_at.get_last_error() != NSAPI_ERROR_OK) {
            return NSAPI_ERROR_AUTH_FAILURE;
        }
    }

    return NSAPI_ERROR_OK;
}

nsapi_error_t AT_CellularNetwork::activate_context()
{
    _at.lock();

    nsapi_error_t err = NSAPI_ERROR_OK;

    // try to find or create context with suitable stack
    if(get_context()) {
        // try to authenticate user before activating or modifying context
        err = do_user_authentication();
    } else {
        err = NSAPI_ERROR_NO_CONNECTION;
    }

   if (err != NSAPI_ERROR_OK) {
.....
```

ytsuboi replied:
@mirelachirica @AnttiKauppila I also needed to change accessibility of get_context(). Applied suggested changes.

AnttiKauppila reviewed (changes requested):
Please modify AT_CellularNetwork class as I suggested in my comment

ytsuboi committed:
[Wio 3G] Changed suggested points

Changed accessibility cellular features member functions, fixed minor target issues

AnttiKauppila commented on features/cellular/framework/AT/AT_CellularNetwork.cpp:
```diff
@@ -420,7 +429,7 @@ void AT_CellularNetwork::ppp_status_cb(nsapi_event_t event, intptr_t parameter)
+nsapi_error_t AT_CellularNetwork::do_user_authentication()
 {
     // try to find or create context with suitable stack
     if (!get_context()) {
```
This check is not needed anymore, it is done in activate_context already

ytsuboi replied:
Done

AnttiKauppila commented on features/cellular/framework/AT/AT_CellularNetwork.h:
```diff
@@ -180,6 +178,8 @@ class AT_CellularNetwork : public CellularNetwork, public AT_CellularBase
     AuthenticationType _authentication_type;
     int _cell_id;
     nsapi_connection_status_t _connect_status;
+    virtual nsapi_error_t do_user_authentication();
```
No need for error value anymore

mirelachirica replied:
Do you mean no error return value? What about NSAPI_ERROR_AUTH_FAILURE?

ytsuboi replied:
I think this topic is not about adding target, so I didn’t make any change about this. Please avoid this topic in this PR.

AnttiKauppila commented on features/cellular/framework/targets/QUECTEL/UG96/QUECTEL_UG96_CellularNetwork.cpp:
```diff
@@ -42,3 +42,29 @@ nsapi_error_t QUECTEL_UG96_CellularNetwork::set_access_technology_impl(RadioAcce
+nsapi_error_t QUECTEL_UG96_CellularNetwork::do_user_authentication()
+{
+    // try to find or create context with suitable stack
+    if (!get_context()) {
```
This check is not needed anymore, it is done in activate_context already

ytsuboi replied:
Done

AnttiKauppila reviewed (changes requested):
do_user_authentication does not need return value anymore, please remove that

AnttiKauppila commented on features/cellular/framework/AT/AT_CellularNetwork.h:
```diff
@@ -180,6 +178,8 @@ class AT_CellularNetwork : public CellularNetwork, public AT_CellularBase
     int _cell_id;
     nsapi_connection_status_t _connect_status;
+    virtual nsapi_error_t do_user_authentication();
+    bool get_context();
```
I don't see a reason why you changed the visibility of this method. Could you explain?

ytsuboi replied:
get_context() was called from do_user_authentication(), so there were compile error. This is why I made change.
```
[Error] AT_CellularNetwork.h@157,10: 'bool mbed::AT_CellularNetwork::get_context()' is private
[Error] QUECTEL_UG96_CellularNetwork.cpp@49,10: within this context
[Error] AT_CellularNetwork.h@157,10: 'bool mbed::AT_CellularNetwork::get_context()' is private
[Error] QUECTEL_UG96_CellularNetwork.cpp@49,22: within this context
[ERROR] In file included from ./mbed-os/features/cellular/framework/targets/QUECTEL/UG96/QUECTEL_UG96_CellularNetwork.h:21:0,
                 from ./mbed-os/features/cellular/framework/targets/QUECTEL/UG96/QUECTEL_UG96_CellularNetwork.cpp:18:
./mbed-os/features/cellular/framework/AT/AT_CellularNetwork.h: In member function 'virtual nsapi_error_t mbed::QUECTEL_UG96_CellularNetwork::do_user_authentication()':
./mbed-os/features/cellular/framework/AT/AT_CellularNetwork.h:157:10: error: 'bool mbed::AT_CellularNetwork::get_context()' is private
     bool get_context();
          ^
./mbed-os/features/cellular/framework/targets/QUECTEL/UG96/QUECTEL_UG96_CellularNetwork.cpp:49:10: error: within this context
     if (!get_context()) {
          ^
In file included from ./mbed-os/features/cellular/framework/targets/QUECTEL/UG96/QUECTEL_UG96_CellularNetwork.h:21:0,
                 from ./mbed-os/features/cellular/framework/targets/QUECTEL/UG96/QUECTEL_UG96_CellularNetwork.cpp:18:
./mbed-os/features/cellular/framework/AT/AT_CellularNetwork.h:157:10: error: 'bool mbed::AT_CellularNetwork::get_context()' is private
     bool get_context();
          ^
./mbed-os/features/cellular/framework/targets/QUECTEL/UG96/QUECTEL_UG96_CellularNetwork.cpp:49:22: error: within this context
     if (!get_context()) {
                      ^
```
So I'm guessing I can return get_context() to private, if I remove [this](https://github.com/ARMmbed/mbed-os/pull/7098#pullrequestreview-127371674) check.

ytsuboi replied:
Now, get_context() backed to private.

ytsuboi committed:
[Wio 3G] Changed suggested things

Removed context check from do_user_authentication.

AnttiKauppila commented:
0xc0170 This is fine from Cellular team perspective, ready to be tested

cmonr commented:
/morph build

cmonr commented:
Failed test does not appear to be caused by the PR, and mirrors an issue we observed earlier today.
/morph test
